File: src/exercise-1.23/answer-1.23.c
/* C solution to SICP exercise 1.23 */

#include "../minunit.h"

int next_divisor(n) {
  return (n == 2) ? 3 : n + 2;
}


int square(int x) {
  return x * x;
}


int divides (int divisor, int n) {
  return ( (n % divisor) == 0);
}


unsigned long smallest_divisor(int n) {
  int divisor = 2;

  for (;;) {
    if (square(divisor) > n)
      return n;
    else if (divides(divisor, n))
      return divisor;
    else
      divisor = next_divisor(divisor);
  }
}


void test_setup() {}
void test_teardown() {}

MU_TEST(test_next_divisor) {
  mu_check(next_divisor(2) == 3);
  mu_check(next_divisor(3) == 5);
  mu_check(next_divisor(5) == 7);
  mu_check(next_divisor(7) == 9);
}


MU_TEST(test_square) {
  mu_check(square(0) == 0);
  mu_check(square(1) == 1);
  mu_check(square(2) == 4);
  mu_check(square(3) == 9);
  mu_check(square(4) == 16);
  mu_check(square(5) == 25);
}


MU_TEST(test_divides) {
  mu_check(! divides(3, 4));
  mu_check(divides(2, 4));
}


MU_TEST(test_smallest_divisor) {
  mu_check(smallest_divisor(199)   == 199);
  mu_check(smallest_divisor(1999)  == 1999);
  mu_check(smallest_divisor(19999) == 7);
}


MU_TEST_SUITE(test_suite) {
  MU_SUITE_CONFIGURE(&test_setup, &test_teardown);
  MU_RUN_TEST(test_next_divisor);
  MU_RUN_TEST(test_square);
  MU_RUN_TEST(test_divides);
  MU_RUN_TEST(test_smallest_divisor);
}


int main (int argc, char *argv[]) {
  MU_RUN_SUITE(test_suite);
  MU_REPORT();
  return 0;
}

File: src/exercise-1.23/answer-1.23.rb
#!/usr/bin/ruby -w

=begin
Ruby solution to SICP exercise 1.23
=end

require 'test/unit'

def next_divisor(n)
  (n == 2) ? 3 : n + 2
end

def square(x)
  x * x
end

def divides(divisor, n)
  ( (n % divisor) == 0)
end

def smallest_divisor(n)
  divisor = 2

  loop do
    if square(divisor) > n
      return n
    elsif divides(divisor, n)
      return divisor
    else
      divisor = next_divisor(divisor)
    end
  end
end


class TestDivisor < Test::Unit::TestCase
  def test_next_divisor
    assert_equal(3, next_divisor(2))
    assert_equal(5, next_divisor(3))
    assert_equal(7, next_divisor(5))
    assert_equal(9, next_divisor(7))
  end

  def test_square
    assert_equal(0,  square(0))
    assert_equal(1,  square(1))
    assert_equal(4,  square(2))
    assert_equal(9,  square(3))
    assert_equal(16, square(4))
    assert_equal(25, square(5))
  end

  def test_divides
    assert(! divides(2, 7))
    assert(divides(2, 6))
  end

  def test_smallest_divisor
    assert_equal(199,  smallest_divisor(199))
    assert_equal(1999, smallest_divisor(1999))
    assert_equal(7,    smallest_divisor(19999))
  end
end

